light/beacon/api/light_api.go

```diff
@@ -0,0 +1,356 @@
+}
+
+// serialize returns the serialized version of the committee
+func (s *syncCommitteeJson) serialize() ([]byte, bool) {
+	if len(s.Pubkeys) != 512 {
+		return nil, false
+	}
+	sk := make([]byte, 513*48)
+	for i, key := range s.Pubkeys {
+		if len(key) != 48 {
+			return nil, false
+		}
+		copy(sk[i*48:(i+1)*48], key[:])
+	}
+	if len(s.Aggregate) != 48 {
+		return nil, false
+	}
+	copy(sk[512*48:], s.Aggregate[:])
+	return sk, true
+}
+
+// GetHead fetches and validates the beacon header with the given blockRoot.
+// If blockRoot is null hash then the latest head header is fetched.
+func (api *BeaconLightApi) GetHeader(blockRoot common.Hash) (beacon.Header, error) {
+	path := "/eth/v1/beacon/headers/"
+	if blockRoot == (common.Hash{}) {
+		path += "head"
+	} else {
+		path += blockRoot.Hex()
+	}
+	resp, err := api.httpGet(path)
+	if err != nil {
+		return beacon.Header{}, err
+	}
+
+	var data struct {
+		Data struct {
+			Root      common.Hash `json:"root"`
+			Canonical bool        `json:"canonical"`
+			Header    struct {
+				Message   beacon.Header `json:"message"`
+				Signature hexutil.Bytes `json:"signature"`
+			} `json:"header"`
+		} `json:"data"`
+	}
+	if err := json.Unmarshal(resp, &data); err != nil {
+		return beacon.Header{}, err
+	}
+	header := data.Data.Header.Message
+	if blockRoot == (common.Hash{}) {
+		blockRoot = data.Data.Root
+	}
+	if header.Hash() != blockRoot {
+		return beacon.Header{}, errors.New("retrieved beacon header root does not match")
+	}
+	return header, nil
+}
+
+// GetStateProof fetches and validates a Merkle proof for the specified parts of the recent
+// beacon state referenced by stateRoot. If successful the returned multiproof has the format
+// specified by expFormat. The state subset specified by the list of string keys (paths) should
+// cover the subset specified by expFormat.
+func (api *BeaconLightApi) GetStateProof(stateRoot common.Hash, paths []string, expFormat beacon.ProofFormat) (beacon.MultiProof, error) {
+	path := "/eth/v1/beacon/light_client/proof/" + stateRoot.Hex() + "?paths=" + paths[0]
+	for i := 1; i < len(paths); i++ {
+		path += "&paths=" + paths[i]
+	}
+	resp, err := api.httpGet(path)
+	if err != nil {
+		return beacon.MultiProof{}, err
+	}
+	proof, err := beacon.ParseSSZMultiProof(resp)
+	if err != nil {
+		return beacon.MultiProof{}, err
+	}
+	var values beacon.MerkleValues
+	reader := proof.Reader(nil)
+	root, ok := beacon.TraverseProof(reader, beacon.NewMultiProofWriter(expFormat, &values, nil))
+	if !ok || !reader.Finished() || root != stateRoot {
+		return beacon.MultiProof{}, errors.New("Invalid state proof")
+	}
+	return beacon.MultiProof{Format: expFormat, Values: values}, nil
+}
+
+// GetCheckpointData fetches and validates bootstrap data belonging to the given checkpoint.
+func (api *BeaconLightApi) GetCheckpointData(ctx context.Context, checkpoint common.Hash) (beacon.Header, beacon.CheckpointData, []byte, error) {
```

Comment:
How can we verify the retrieved checkpoint is matched with the provided checkpointHash?

We only verify the retrieved committee is indeed matched with the associated header in the RPC response, but the entire response is not verified with the provided checkpointHash?

Reply:
Indeed, good catch, I forgot to add this check.